File: jphoenix-emulator-port/MameLofiResampler.java
final class MameLofiResampler {
	interface Source {
		double nextSample();
	}

	static final int PHASE_ONE = 0x1000000;
	static final int PHASE_MASK = PHASE_ONE - 1;
	private static final float[][] INTERPOLATION = createInterpolation();

	private final int sourceDivide;
	private final int step;
	private int phase;
	private float s0;
	private float s1;
	private float s2;
	private float s3;

	MameLofiResampler(int sourceRate, int targetRate) {
		if (sourceRate <= 0 || targetRate <= 0) {
			throw new IllegalArgumentException("Sample rates must be positive");
		}
		sourceDivide = sourceRate <= targetRate ? 1 : 1 + sourceRate / targetRate;
		step = (int) ((long) sourceRate * PHASE_ONE / targetRate / sourceDivide);
	}

	double nextSample(Source source) {
		int currentPhase = phase >>> 12;
		float output =
				-s0 * INTERPOLATION[0][0x1000 - currentPhase]
				+ s1 * INTERPOLATION[1][0x1000 - currentPhase]
				+ s2 * INTERPOLATION[1][currentPhase]
				- s3 * INTERPOLATION[0][currentPhase];

		phase += step;
		if ((phase & PHASE_ONE) != 0) {
			phase &= PHASE_MASK;
			s0 = s1;
			s1 = s2;
			s2 = s3;
			s3 = readSource(source);
		}
		return output;
	}

	private float readSource(Source source) {
		float sum = 0.0f;
		for (int i = 0; i < sourceDivide; i++) {
			sum += (float) source.nextSample();
		}
		return sum / sourceDivide;
	}

	int sourceDivide() {
		return sourceDivide;
	}

	int step() {
		return step;
	}

	void writeState(java.io.DataOutput output) throws java.io.IOException {
		output.writeInt(sourceDivide);
		output.writeInt(step);
		output.writeInt(phase);
		output.writeFloat(s0);
		output.writeFloat(s1);
		output.writeFloat(s2);
		output.writeFloat(s3);
	}

	void readState(java.io.DataInput input) throws java.io.IOException {
		int savedSourceDivide = input.readInt();
		int savedStep = input.readInt();
		if (savedSourceDivide != sourceDivide || savedStep != step) {
			throw new java.io.IOException("audio resampler configuration mismatch");
		}
		phase = input.readInt();
		if ((phase & ~PHASE_MASK) != 0) {
			throw new java.io.IOException("invalid audio resampler phase");
		}
		s0 = input.readFloat();
		s1 = input.readFloat();
		s2 = input.readFloat();
		s3 = input.readFloat();
	}

	static float interpolation(int lane, int index) {
		return INTERPOLATION[lane][index];
	}

	private static float[][] createInterpolation() {
		float[][] table = new float[2][0x1001];
		for (int i = 1; i < 4096; i++) {
			float p = i / 4096.0f;
			table[0][i] = (p - p * p * p) / 6.0f;
		}
		for (int i = 1; i < 2049; i++) {
			float p = i / 4096.0f;
			table[1][i] = p + (p * p - p * p * p) / 2.0f;
		}
		for (int i = 2049; i < 4096; i++) {
			table[1][i] = 1.0f + table[0][i] + table[0][4096 - i] - table[1][4096 - i];
		}
		table[0][0] = 0.0f;
		table[0][0x1000] = 0.0f;
		table[1][0] = 0.0f;
		table[1][0x1000] = 1.0f;
		return table;
	}
}

File: redot-port/native/phoenix-core/mame_lofi_resampler.c
#include "mame_lofi_resampler.h"
#include <stdbool.h>

#define PHASE_ONE 0x1000000
#define PHASE_MASK (PHASE_ONE - 1)

static float g_interp[2][0x1001];
static bool g_interp_ready = false;

/*
 * Builds the cubic interpolation lookup tables used by MAME's lo-fi stream
 * resampler. The tables are process-global because the coefficients depend
 * only on fractional phase, not on the source stream.
 */
static void build_interpolation(void) {
    if (g_interp_ready) return;
    for (int i = 1; i < 4096; i++) {
        float p = i / 4096.0f;
        g_interp[0][i] = (p - p * p * p) / 6.0f;
    }
    for (int i = 1; i < 2049; i++) {
        float p = i / 4096.0f;
        g_interp[1][i] = p + (p * p - p * p * p) / 2.0f;
    }
    for (int i = 2049; i < 4096; i++) {
        g_interp[1][i] = 1.0f + g_interp[0][i] + g_interp[0][4096 - i] - g_interp[1][4096 - i];
    }
    g_interp[0][0] = 0.0f;
    g_interp[0][0x1000] = 0.0f;
    g_interp[1][0] = 0.0f;
    g_interp[1][0x1000] = 1.0f;
    g_interp_ready = true;
}

/*
 * Initializes one rate converter from a native generator rate to the SDL
 * output rate. source_divide folds very high native rates down before the
 * interpolator so callers can use the same path for TMS and discrete sound.
 */
void mame_lofi_resampler_init(MameLofiResampler* r, int source_rate, int target_rate) {
    build_interpolation();
    r->source_divide = (source_rate <= target_rate) ? 1 : (1 + source_rate / target_rate);
    r->step = (int)((long long)source_rate * PHASE_ONE / target_rate / r->source_divide);
    r->phase = 0;
    r->s0 = r->s1 = r->s2 = r->s3 = 0.0f;
}

/*
 * Reads and averages the required number of native samples for the next source
 * point. Averaging here preserves the original MAME-style downsampling shape
 * without exposing source_divide to individual sound generators.
 */
static float read_source(MameLofiResampler* r, MameLofiSource source, void* ctx) {
    float sum = 0.0f;
    for (int i = 0; i < r->source_divide; i++) {
        sum += (float)source(ctx);
    }
    return sum / r->source_divide;
}

/*
 * Produces one target-rate sample from the four-sample interpolation window.
 * When fixed-point phase crosses a source-sample boundary, the window advances
 * and pulls the next averaged native sample from the callback.
 */
double mame_lofi_resampler_next(MameLofiResampler* r, MameLofiSource source, void* ctx) {
    int current_phase = (unsigned int)r->phase >> 12;
    float output = -r->s0 * g_interp[0][0x1000 - current_phase]
                 +  r->s1 * g_interp[1][0x1000 - current_phase]
                 +  r->s2 * g_interp[1][current_phase]
                 -  r->s3 * g_interp[0][current_phase];

    r->phase += r->step;
    if (r->phase & PHASE_ONE) {
        r->phase &= PHASE_MASK;
        r->s0 = r->s1;
        r->s1 = r->s2;
        r->s2 = r->s3;
        r->s3 = read_source(r, source, ctx);
    }
    return output;
}
